# Домашнє завдання до теми "Розширені методи обробки даних у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import numpy as np
import pandas as pd

In [3]:
data_path = '../data/supermarket_sales.csv'
df = pd.read_csv(data_path)
type(df)

pandas.core.frame.DataFrame

1. Відсортуйте дані за спаданням ціни товара ('Unit price') і виведіть 5 перших рядочків на екран.

In [4]:
df.sort_values(by="Unit price", ascending=False).head(5)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
983,148-41-7930,C,Naypyitaw,Normal,Male,Health and beauty,99.96,7,34.986,734.706,1/23/2019,10:33,Cash,699.72,4.761905,34.986,6.1
122,219-22-9386,B,Mandalay,Member,Male,Sports and travel,99.96,9,44.982,944.622,3/9/2019,17:26,Credit card,899.64,4.761905,44.982,4.2
930,641-62-7288,B,Mandalay,Normal,Male,Home and lifestyle,99.92,6,29.976,629.496,3/24/2019,13:33,Ewallet,599.52,4.761905,29.976,7.1
494,437-53-3084,B,Mandalay,Normal,Male,Fashion accessories,99.89,2,9.989,209.769,2/26/2019,11:48,Ewallet,199.78,4.761905,9.989,7.1
283,667-92-0055,A,Yangon,Member,Male,Health and beauty,99.83,6,29.949,628.929,3/4/2019,15:02,Ewallet,598.98,4.761905,29.949,8.5


2. Знайдіть інвойс (рядок в даних) в категорії товарів `Electronic accessories`, в якому був куплений товар з найбільшою кількістю і найбільшою ціною. Виведіть цей рядок даних на екран.

In [5]:
df[df['Product line'] == 'Electronic accessories'] \
    .sort_values(['Quantity', 'Unit price'], ascending=False) \
    .head(1)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
105,704-48-3927,A,Yangon,Member,Male,Electronic accessories,88.67,10,44.335,931.035,1/12/2019,14:50,Ewallet,886.7,4.761905,44.335,7.3


3.1. Створіть нову колонку в даних `Rating Category`, яка визначає категорію рейтингу (`Rating`): високий, середній чи низький за наступним правилом:
- якщо рейтинг вищий або рівний 8 - то він високий (`High`)
- якщо рейтинг від 6 (включно) до 8 (не включно) - то він середній (`Medium`)
- інакше - низький (`Low`).

In [6]:
def rating_category(r):
    if r >= 8:
        return "High"
    elif r >= 6:
        return "Medium"
    else:
        return "Low"

df['Rating Category'] = df['Rating'].apply(rating_category)


In [8]:
df[['Rating', 'Rating Category']].head(10)

,Rating,Rating Category
0,9.1,High
1,9.6,High
2,7.4,Medium
3,8.4,High
4,5.3,Low
5,4.1,Low
6,5.8,Low
7,8.0,High
8,7.2,Medium
9,5.9,Low


3.2. Виведіть, який відсоток даних становить кожне зі значень категорій рейтингу.

In [10]:
df['Rating Category'].value_counts(normalize=True) 

Rating Category
Medium    0.356
High      0.329
Low       0.315
Name: proportion, dtype: float64

4.1. Запустіть код нижче аби створити змінну `time`. Ця змінна містить час в форматі "година:хвилина". Напишіть функцію `get_hour`, яка буде зі змінної формату як `time` дістати лише години і конвертувати їх у тип `int`.

Запустіть написану функцію на змінній `time` та перевірте правильність роботи функції.

In [11]:
time = '11:28'

In [12]:
def get_hour(t):
    return int(t.split(':')[0])


In [15]:
type(get_hour(time))

int

In [16]:
get_hour(time)

11

4.2. З допомогою функції `get_hour` і метода `DataFrame.apply()` створіть нову колонку `Hour` в нашому наборі даних, яка буде містити години здійснення покупки, обчислені на основі часу здійснення покупки `Time`.

In [17]:
df['Hour'] = df['Time'].apply(get_hour)

In [18]:
df[['Time', 'Hour']].head(5)

,Time,Hour
0,13:08,13
1,10:29,10
2,13:23,13
3,20:33,20
4,10:37,10


4.3. Виведіть кількість значень за кожною годиною покупки і дайте відповідь "О яких годинах найбільше здійснюють покупок покупці і чому на Ваш погляд саме о цих годинах? Чи це слогічно, що найбільше покупок саме в ці години?".

In [19]:
df['Hour'].value_counts().sort_index()

Hour
10    101
11     90
12     89
13    103
14     83
15    102
16     77
17     74
18     93
19    113
20     75
Name: count, dtype: int64

Найбільше покупок здійснюють о **19:00** (113 інвойсів). Також високими за активністю є години **10:00, 13:00** та **15:00**.

Це логічно, оскільки о 19:00 більшість покупців повертаються з роботи й заходять до магазинів дорогою додому. Обідній час (13:00) теж має високий попит, бо люди роблять покупки під час перерви. Ранній день (10:00) та середина дня (15:00) також є типовими періодами зростання активності в роздрібній торгівлі.

Отже, пікові години повністю відповідають нормальній поведінці покупців.

5. Виведіть середнє значення чеку (`Total`) для покупок в містах, наявних в даних, за допомогою функції `groupby`.

In [21]:
df.groupby('City')['Total'].mean().round(2)

City
Mandalay     319.87
Naypyitaw    337.10
Yangon       312.35
Name: Total, dtype: float64

6. Виведіть кількість покупок, середнє, мінімальне та максимальне значення для загальної суми чеку (Total) в містах, наявних в даних, за допомогою функції groupby.
Зробіть висновки з цих показників та дайте рекомендації для мережі супермаркетів.

In [22]:
result = (
    df.groupby('City')['Total']
      .agg(['count', 'mean', 'min', 'max'])
      .round(2)
)

result


,count,mean,min,max
City,,,,
Mandalay,332,319.87,18.64,1022.49
Naypyitaw,328,337.10,10.68,1042.65
Yangon,340,312.35,12.69,1039.29


### Висновки 

На основі агрегованої статистики видно, що міста мають суттєві відмінності в активності покупців і середньому чекові:

1. **Кількість покупок (count)** показує, наскільки активна торгівля в кожному місті. Місто з найбільшою кількістю інвойсів забезпечує найбільший трафік покупців.

2. **Середній чек (mean)** відрізняється між містами. Вищий середній чек може означати:
   - вищу купівельну спроможність населення,
   - більшу популярність дорогих товарних категорій,
   - ефективнішу роботу маркетингу та мерчендайзингу.

3. **Мінімальні та максимальні значення Total (min, max)** дають уявлення про різноманітність покупок у місті:
   - низький мінімум означає покупки дрібних товарів,
   - високий максимум — наявність великих чеків.

### Рекомендації 

1. **Посилити маркетинг у містах з нижчим середнім чеком**, зробивши акцент на:
   - промо-акціях,
   - бонусних програмах,
   - апсейлі (додавання супутніх товарів).

2. **У містах з найбільшим середнім чеком посилити акцент на преміальні товари та категорії**, оскільки покупці вже демонструють більшу готовність витрачати.

3. **У місті з найбільшою кількістю покупок варто збільшити персонал у пікові години**, покращити логістику викладки та контроль запасів.

4. **Аналізувати категорії товарів**, які формують найбільший Total у кожному місті, та оптимізувати їх наявність.

5. **Запустити локальні маркетингові кампанії**, адаптовані під купівельну поведінку кожного міста.



7. Ми внесли деякі зміни в наш набір даних, тож, давайте запишемо результат у файл. Запишіть оновлену версію даних у файл `supermarket_sales_enriched.csv`. Зверніть увагу, аби не записати нічого зайвого, наприклад, індекси рядків нам в файлі не потрібні, оскільки вони стандартні.

In [23]:
df.to_csv('supermarket_sales_new.csv', index=False)

In [25]:
pd.read_csv('supermarket_sales_new.csv', nrows=5)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating,Rating Category,Hour
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1,High,13
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6,High,10
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4,Medium,13
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4,High,20
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3,Low,10
